org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/WithServiceTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit5.service;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.common.annotation.ServiceProperty;
+import org.osgi.test.common.annotation.WithService;
+import org.osgi.test.junit5.service.WithServiceTest.MyFoo;
+import org.osgi.test.junit5.types.Foo;
+
+@ExtendWith(ServiceExtension.class)
+@WithService(ofClass = MyFoo.class, value = Foo.class, properties = {
```

Comment:
The whole concept seems extremely limited. Since you can only use it for classes having a public, no-arg constructor. Or is the some other way to make a service object that you are not showing/testing here? I think the lack of javadoc on WithService is leaving me in the dark.

Reply:
You cann call a static method inside the testclass to get an object.  Will add javadoc

Reply:
And tests. You may want to look at org.junit.jupiter.params.provider.MethodSource for inspiration here.

Reply:
Might be nice if it could also call instance methods rather than just static methods - would increase the possible number of use cases. The factory method in this case could, for example, create a service implementation using Mockito and store a reference to it in a test class instance field - then you can perform tests on it using Mockito.verify(). Although, I do wonder if there is much ground gained here over (eg) simply calling bundleContext.registerService() at the start of your test method.

Reply:
For me the cleanup (unregister) and annotating nestes classes is the interesting part.

Reply:
Clean up of registered services is already done by the BundleContextExtension. So you can just use it to get a BundleContext and then use the normal BundleContext.registerService method. So @kriegfrj's comment on just using BundleContext.registerService is quite valid.

Reply:
> Will have a look how to get the TestInstance to call more than static methods. in my test i just got null as instance.

`@WithService` tags that are registered at the class level, it would make sense to inject them in `ServiceExtension.beforeAll()`. If your class is running with `TestLifecycle.PER_CLASS`, then the extension context will have a non-null test instance at this stage; otherwise you'd throw an `ExtensionConfigurationException` if you're trying to use a non-static initialiser method (similar to how Jupiter throws an exception if it binds a non-static `@BeforeAll` method on a `PER_CLASS` annotated test class).

For `@WithService` tags registered on an individual test method, you'd inject them in `ServiceExtension.beforeEach()`. The extension context will always have a test class instance available in the beforeEach() callback regardless of the lifecycle mode you're using.
For services register

> For me the cleanup (unregister) and annotating nestes classes is the interesting part.
The unregister is already handled by `BundleContextExtension` if you use `registerService()` through the injected `BundleContext` though. What I wrote above could also be achieved simply by calling `registerService()` in a `@BeforeAll` method. Hence why I see the gain of this annotation as minimal.

If you could elaborate on the "annotating nested classes" part though, perhaps you could enlighten me as to where it starts to get more interesting.

Reply:
> `@WithService` tags that are registered at the class level, it would make sense to inject them in `ServiceExtension.beforeAll()`.

This comment applies to nested classes too.

Reply:
> Clean up of registered services is already done by the BundleContextExtension. So you can just use it to get a BundleContext and then use the normal BundleContext.registerService method. So @kriegfrj's comment on just using BundleContext.registerService is quite valid.

@stbischof , perhaps given the above comment, we should try and more fully thrash out what the potential gain might be of this annotation before you invest too much effort polishing it up. Some other possible arguments for this approach:

- Declarative configuration (configuration-by-annotation) could be considered "neater" than configuration-in-code
- Configuring the properties dictionary declaratively is somewhat more convenient than building a hashtable in code
- No need to explicitly inject the bundle context. Often you will need the injected BC for other reasons, so you might not save anything here. But there will be cases where you might get away with not injecting the BC at all by using this annotation instead.
- You could allow combining `@WithService` with `@InjectService` functionality, adopting the pattern that `ConfigurationExtension` uses. Makes for some nice symmetry.

On the other hand, I'm not sure that the initialiser method approach gains us much. If you need explicit code to create the service, perhaps you're just as well off to register your service in-code using the injected BundleContext?

At this stage I'm still not 100% sold for or against, though I think having enumerated some of these possible gains I'm leaning towards "for" implementing `@WithService` in some form.

Reply:
I see the point. At the moment i see much more benefits and less written code on the bundleContext side. And you are more flexible if you wand to create new a service of the same class '@beforeEach' Testmethod.

Reply:
I think the deal breaker here is that most service instances will need more than a public, no-args constructor to create the service object. So you will always need to run some code to construct the service object. So there is not much to be gained by using an annotation to only handle the service registration if you also must author a service object construction method.

Reply:
There is still some value in this PR. I like the rename of ConfigEntry to Property and the related changes. So we can at least keep that.

Reply:
what do you think about TargetType.java?

Reply:
> what do you think about TargetType.java?

That also looks useful to keep.

Reply:
> I think the deal breaker here is that most service instances will need more than a public, no-args constructor to create the service object. So you will always need to run some code to construct the service object. So there is not much to be gained by using an annotation to only handle the service registration if you also must author a service object construction method.

One special case that I can see is when you'd like to register a mock service implementation. Perhaps a `@WithMockService` implementation that just takes the service type. Then behind the scenes it takes care of calling `mock()` to give you a mock object you can then verify against. Useful for testing client code of simple, stateless services.

Reply:
> @WithMockService

Seems rather highly specialized and would imply an implementation coupling to mockito which I would prefer we avoid.

Reply:
sticking my nose in late but I think I agree that there is no good general solution here. I think publishing services with any of the lifecycle scope methods is really the most practical approach, particularly because you already benefit from cleanup garnered by the BundleContext (you _should_ be using the bundle context supplied by the extension, right?)

Reply:
> you _should_ be using the bundle context supplied by the extension, right?)

Yes, you always want to use the BundleContext supplied by BundleContextExtension since it cleans up with proper scoping.

Reply:
> > @WithMockService
> 
> Seems rather highly specialized and would imply an implementation coupling to mockito which I would prefer we avoid.

Yes, i would want to avoid tight coupling to Mockito too - at least in the core modules. However,  I'm open to the idea of a mockito module which contained Mockito-specific utilities v such as this. I thi